Issue #1013. Fix NoWhitespaceAfterCheck

attatrol opened the pull request:
It satisfies issue #1013

1. Behavior of NoWhitespaceAfterCheck for specific token type <code>TokenTypes.ARRAY_DECLARATOR</code> is overhauled.
Now check generates error if there are whitespaces in highlighted places:
<code>
int█[ ]█[ ] a;</code>
<code>int a█[ ]█[ ];</code>
<code>String█[ ]█[ ] someStuff6 = (java.lang.String█[ ]█[ ]) someStuff5;</code>
<code>return new Integer█[  i + j  ]█[i + j]█[0];
</code>
it does not check for whitespace validity inside brackets.

2. Token <code>TokenTypes.INDEX_OP</code> is added to the lists of default and accepted tokens due to its simularity with <code>TokenTypes.ARRAY_DECLARATOR</code>.

3. Other token types processing remains intact.

4. There is an error in Travis-CI build job no.8:
<code>[ERROR] src/main/java/org/orekit/propagation/semianalytical/dsst/forces/DSSTThirdBody.java[1087:77] (whitespace) NoWhitespaceAfter: 'fact' is followed by whitespace.</code>
https://travis-ci.org/checkstyle/checkstyle/jobs/72631540

It happens because updated check finds waste whitespace in Orecit source:
https://www.orekit.org/forge/projects/orekit/repository/revisions/master/entry/src/main/java/org/orekit/propagation/semianalytical/dsst/forces/DSSTThirdBody.java

romani commented:
Error: 
[ERROR] src/main/java/org/orekit/propagation/semianalytical/dsst/forces/DSSTThirdBody.java[1087:77] (whitespace) NoWhitespaceAfter: 'fact' is followed by whitespace.

please fix a problem by PR to our fork of that project - https://github.com/checkstyle/Orekit
After that I will relaunch Travis.

attatrol commented:
Done
https://github.com/checkstyle/Orekit/pull/3

romani commented:
@attatrol , please rebase your changes, to resolve conflict, sorry that I merged changes of others first.

romani commented:
please confirm that after reverting https://github.com/CS-SI/Orekit/commit/05c49111b549a2226ee899d664b6f3d001439441 there is no violations on the code, to prove that regression that we introduced between 6.2 and 6.3 is eliminated, please recheck that on your local and then send to checkstyle/Orekit revert of this commit to prove that problem is resolved and author of CS-SI/Orekit can revert that change too after our release (6.9 is planned for 30 july).

romani commented:
1) 
> // cobertura hates useful dead code!

please remove emotions from code and put a explanation

2)
you did too much changes to code, please rebase  code for now, please expect that merge will take a time as I need to be sure that we are not going to fix one regression and introduce another. Please think how you can prove that your code works well to speed up acceptance process. Please look at others PRs how that is done. 
Please review https://github.com/checkstyle/contribution/tree/master/checkstyle-tester simple shell script that will help you test a change of huge sources.

attatrol commented:
<blockquote>
please confirm that after reverting CS-SI/Orekit@05c4911 there is no violations on the code, to prove that regression that we introduced between 6.2 and 6.3 is eliminated, please recheck that on your local and then send to checkstyle/Orekit revert of this commit to prove that problem is resolved and author of CS-SI/Orekit can revert that change too after our release (6.9 is planned for 30 july).
</blockquote>
<p>
I modified travis.yml to use my local copy of Orecit, reverted subject commit, got 10 warnings:
https://travis-ci.org/attatrol/checkstyle/jobs/72796539
All of this warnings fit into new processing of <code>TokenTypes.ARRAY_DECLARATOR</code>.
They were corrected and travis rerun with positive outcome.
https://travis-ci.org/attatrol/checkstyle/builds/72800414
The PR for CS-SI/Orekit was staged:
https://github.com/checkstyle/Orekit/pull/4
</p>

1)
<blockquote>
please remove emotions from code and put a explanation
</blockquote>
<p>
I apologize for this, i was puzzled of how to reach 100% UT coverage for a dead code. I've learned about mocks.
</p>

2) - 6) 
Done.

romani commented:
> The PR for checkstyle/Orekit was staged:

I applied it but ,now cause problems for all other PRs, I am doing revert of them.

romani commented:
@attatrol , lets focus on that PR, it is hanging too much

attatrol commented:
@romani,
What do you mean, this PR or NPath complexity issue?

romani commented:
1)
> parentTypeTypeProcessing
> getArrayDeclaratorPreviousElement

please make it as single exit point method.

2)
>  ident.getLineNo() != ast.getLineNo() ? ident.getLineNo() > ast.getLineNo()

yes that is cool construction , but please change it traditional IF-ELSE , as it is not human friendly, next developer after you will pause there for a while to understand it. "Code need to be read by diagonal like a text".

attatrol commented:
1), 2)
Done.

romani commented:
1)
> parentTypeTypeProcessing

please fix typo.

2) Changes are huge for a Check, It is very hard to be sure that did not introduce any regression.
So lets do that by testing. Please use checkstyle-tester not only for guava but for all other projects (without ARRAY token to minimize violations, as you did it before) - no diff is expected. 
http://checkstyle.sourceforge.net/config_whitespace.html#NoWhitespaceAfter please also do that for option allowLineBreaks.

romani commented:
@attatrol , do you have time to finish that PR ?

attatrol commented:
i'll finish it immediately after summer ends.

2015-08-26 19:09 GMT+03:00 Roman Ivanov <notifications@github.com>:

> @attatrol <https://github.com/attatrol> , do you have time to finish that
> PR ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/checkstyle/checkstyle/pull/1474#issuecomment-135081667>
> .
>

romani commented:
@attatrol , can you share when you will have time to finish that PR ?

mkordas commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheck.java:
```diff
@@ -167,163 +157,203 @@ private static int getPositionAfter(DetailAST ast) {
-     * @return true if current array is used as type for generic bounded wildcard.
+     * Returns proper argument for getPositionAfter method, it is a token after
+     * {@link TokenTypes#ARRAY_DECLARATOR ARRAY_DECLARATOR}, in can be {@link TokenTypes#RBRACK
+     * RBRACK}, {@link TokenTypes#INDENT INDENT} or an array type definition (literal).
```
IDEA Inspections have just detected problem in this line - there's no `TokenTypes.INDENT` available: https://teamcity.jetbrains.com/viewLog.html?buildId=563116&tab=Inspection&buildTypeId=Checkstyle_IdeaInspectionsPullRequest

romani commented:
@attatrol , please share your html reports from checkstyle-tester on github.io for review.

attatrol commented:
<p>
1) <blockquote>
please make it as single exit point method.
</blockquote>
Done.
</p>
<p>
2) <blockquote>
please share your html reports from checkstyle-tester on github.io for review.
</blockquote>
NoWhitespaceAfter check fails with this configuration:
<blockquote>
	 	&lt;module name="NoWhitespaceAfter" &gt;<br>
	 	          &lt;property name="allowLineBreaks" value="false" /&gt;<br>
	 	&lt;/module &gt;<br>
</blockquote>
while checking <a href="http://hg.openjdk.java.net/jdk7/jdk7/jdk/file/9b8c96f96a0f/src/share/classes/javax/print/StreamPrintServiceFactory.java">this file</a> 
both for original and forked versions of the check. Mistake is in erroneous calculation of the string index in <i>getPositionAfter</i> method, which wasn't changed in this PR.<br>
So, only sites with allowLineBreaks=true were generated.<br>
Both sites are over 1Gb each, thanks to <code>TokenTypes.ARRAY_INIT</code> token.
 warnings, now i'm going to host them somewhere, this will consume some time. Or i can create reports without <code>TokenTypes.ARRAY_INIT</code>, because <code>TokenTypes.ARRAY_INIT</code> processing wasn't changed in this PR.

romani commented:
to make teamcity pass , just rebase on latest master

romani commented:
> Both sites are over 1Gb each

please minimize report to reasonable size by tokens or by removing some projects. Report should not be big to let human validate it.

romani commented:
ping.

romani commented:
@attatrol , do you still have time to finish this PR ?

attatrol commented:
I'm neither procrastinating nor quitting, i  have spare time only on weekends because of intensive preparation for major math competition (which is an unpleasant surprise of early autumn). I want to create automatic tool that returns difference between two sites: https://github.com/attatrol/ahsm

romani commented:
release 6.11will be on this Sunday (27th of September), it would be good to finish this PR before release

mkordas commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheck.java:
```diff
@@ -103,17 +104,27 @@
-        return ArrayUtils.EMPTY_INT_ARRAY;
-    }
-
-    @Override
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [The method visitToken() has an NPath complexity of 16](https://www.codacy.com/app/checkstyle/checkstyle/file/1020687841/issues/source?bid=2193633&fileBranchId=2739119#l112)

mkordas commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheck.java:
```diff
@@ -103,17 +104,27 @@
-    @Override
     public void visitToken(DetailAST ast) {
-        final DetailAST astNode = getPreceded(ast);
+        DetailAST astNode;
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [A switch statement does not contain a break](https://www.codacy.com/app/checkstyle/checkstyle/file/1020687841/issues/source?bid=2193633&fileBranchId=2739119#l114)

mkordas commented on src/test/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheckTest.java:
```diff
@@ -104,9 +110,106 @@ public void testArrayDeclarations() throws Exception {
+     * @param tokenColumn token position in a file (column)
+     * @return AST node for the token
+     */
+    private static DetailAST mockAST(final int tokenType, final String tokenText,
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [Avoid long parameter lists.](https://www.codacy.com/app/checkstyle/checkstyle/file/1020688079/issues/source?bid=2193633&fileBranchId=2739119#l203)

romani commented:
@attatrol , are you going to finish this PR ?  It is ok to answer "No". I just do not want to let this PR hang for long time, another contributor can continue your work (the only drawback: you will loose authorship of update after rebase).

mkordas commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheck.java:
```diff
@@ -103,19 +104,27 @@
-        return ArrayUtils.EMPTY_INT_ARRAY;
-    }
-
-    @Override
```
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [The method visitToken() has an NPath complexity of 16](https://www.codacy.com/app/checkstyle/checkstyle/file/1094131143/issues/source?bid=2193633&fileBranchId=2739119#l112)

romani replied:
@mkordas , what do you expect there ? If it is just counting of metrics  - please switch it off. We need notification only if method is more then approved limit.

attatrol commented:
<p>
<blockquote>Codacy Issue found: Avoid long parameter lists.</blockquote>
Fixed.
</p>
<p>
<a href="http://attatrol.github.io/site_result_true/checkstyle_merged.html">Here is a difference between</a> checkstyle-tester site reports for NoWhitespaceAfterCheck before and after PR. That means that identical warnings from both were deleted, and additional row "Site" points to the source of leftover warnings. It was created with <a href="https://github.com/attatrol/ahsm">this unpolished tool</a>. All projects-to-test-on were used.<br>
The check was configured with<code>allowLineBreaks="true"</code>.</p>
<p>
With <code>allowLineBreaks="false"</code> PR passes through tester (and generates enormous site (<a href="http://attatrol.github.io/checkstyle.html">here without source references</a>)) but original check fails with<br><code>java.lang.StringIndexOutOfBoundsException: String index out of range:</code><br> thrown from <br><code>NoWhitespaceAfterCheck.visitToken(NoWhitespaceAfterCheck.java:124)
</code>.<br>

romani commented:
> Here is a difference between checkstyle-tester site reports 

interesting tool ....

if you generate result merged html, please add anchor image like it is done at http://checkstyle.sourceforge.net/google_style.html#a2.3.3 or similar (anchor name could be randomly generated or .... whatever), because during result discussion such direct links to a case are extremely helpful, as they point to exact violation case .

attatrol commented:
<blockquote>please add anchor image</blockquote>
I don't understand what you mean.<br>
Both sites used for merging were generated with this my_check.xml:<br>
<blockquote>
&lt;module name="Checker"&gt;<br>
&lt;property name="charset" value="UTF-8"/&gt;<br>
&lt;property name="severity" value="warning"/&gt;<br>
&lt;module name="TreeWalker"&gt;<br>
&lt;module name="NoWhitespaceAfterCheck"&gt;<br>
&lt;property name="allowLineBreaks" value="true"/&gt;<br>
&lt;/module&gt;<br>
&lt;/module&gt;<br>
&lt;/module&gt;<br>
</blockquote><br>
so i dont see any need in site which lists lots of checks.<p>
<p>

romani commented:
I did not mean multiple Checks. 

I would like to have in a report to have unique anchor (I send you a link as example of how we did this for each table row) for each line in diff-report to let me send it to you and discuss certain violation on certain file and certain  line number.

attatrol commented:
<b>@romani</b>,<br>
anchors added to <a href="http://attatrol.github.io/site_result_true/checkstyle_merged.html">the report.</a>

mkordas commented:
@attatrol, thanks. Can you rebase your commit? We have some conflict here.

romani commented:
@attatrol, thanks a lot. Can you rebase your commit one more time? we have conflict again.
Sorry for caused inconvenience.

romani commented:
Sorry for long delay in code-review, now your are first in my TODO list.

1)
>  final DetailAST nodeBeforeAst = getNodeBefore(ast);

name it getPreviousAst, and variable previousAst.

2)
> private static DetailAST getNodeBefore(DetailAST ast) {

please explain why getNodeBefore return current token by default ? if explanation is short please add it to code as comment.

3)
> public void setAllowLineBreaks(boolean allowLineBreaks) {

please move this method to top before all Override methods. Please recheck that all other methods after Overrides are defined in order of appearance in code above(from top to bottom).

4)
> private boolean hasRedundantWhitespace(String line, int after) {

this method looks strange, can you move out "allowLineBreaks" and make it as findXXXXXXX , you can reuse some guava library methods to avoid iterating by index.


review to be continued ......

attatrol commented:
@romani,
i'm working on fixes but you have some misunderstanding in my report:
<blockquote>
6) false-positive - http://attatrol.github.io/site_result/checkstyle_merged.html#iers.1.97 it is not ideal code but still this not validation of this Check. Multiple declarations and definitions should be ok for this Check.
</blockquote>
Orange coloured lines which have value "site_old" in column "site" are false-positives from previous version of the check, they aren't mentioned in PR report.<br> Green lines are false-negatives from previous version of the check, they are new lines in the PR report.<br> All identical lines of each report are omitted.

attatrol commented:
1). <blockqoute>
    final DetailAST nodeBeforeAst = getNodeBefore(ast);
name it getPreviousAst, and variable previousAst.
    renamed in whitespaceFollowedAst

romani commented:
> but you have some misunderstanding in my report

sorry, my bad, I will continue to review report.

8)
>         if ((after >= line.length() || Character.isWhitespace(line.charAt(after)))
                 && hasRedundantWhitespace(line, after)) {

please make it following, to show priority of execution visually
```
        if ((after >= line.length() 
               || Character.isWhitespace(line.charAt(after))) && hasRedundantWhitespace(line, after)) {
```

8)
>         DetailAST candidate;

please rename to "result" in all use-cases.

9)
>             final int parentType = parent.getType();
            // generics
            if (parentType == TokenTypes.TYPE_ARGUMENT) {

Why not to use a SWITCH there ?

10)
> throw new IllegalStateException("unexpected ast syntax" + parent);

by what Input file you have unexpected syntax ? Checkstyle do not work reliably on non-compilable sources by design.

11)
> private static DetailAST getArrayDeclaratorParent(DetailAST ast) {

rename to getFirstNonArrayDeclaratorParent.

romani commented:
12) 
Please confirm that after reverting https://github.com/CS-SI/Orekit/commit/05c49111b549a2226ee899d664b6f3d001439441 checkstyle validation passed without any problem on this project.

romani commented:
@attatrol , your report is impressive , thanks a lot !!! 
I would love to incorporate your tool in Checkstyle testing. 

All items in report are valid. I am waiting for fixes I described above and we are good to merge.

13) after you rebase your code over our HEAD you should got build failure at http://attatrol.github.io/site_result/checkstyle_merged.html#nput.1.54 , so please provide fix for this case by separate commit of the same PR.

attatrol commented:
<p>
1)
<blockquote>
final DetailAST nodeBeforeAst = getNodeBefore(ast);<br><br>

name it getPreviousAst, and variable previousAst.
</blockquote>
now they are <i>getWhitespaceFollowedAst</i> and <i>whitespaceFollowedAst</i>, see next.
</p>
<p>
2)
<blockquote>
private static DetailAST getNodeBefore(DetailAST ast) {<br><br>

please explain why getNodeBefore return current token by default ? if explanation is short please add it to code as comment.
</blockquote>
It was the most recent refactoring where i lowered complexity of the <i>visitToken</i> and failed to name extracted method properly. Now it is renamed as <i>getWhitespaceFollowedAst</i> because it returns the node which should be checked for trailing whitespace.<br>
</p>
<p>
3)
<blockquote>
public void setAllowLineBreaks(boolean allowLineBreaks) {<br><br>

please move this method to top before all Override methods. Please recheck that all other methods after Overrides are defined in order of appearance in code above(from top to bottom).
</blockquote>
Done.
</p>
<p>

4)
<blockquote>
private boolean hasRedundantWhitespace(String line, int after) {<br><br>

this method looks strange, can you move out "allowLineBreaks" and make it as findXXXXXXX , you can reuse some guava library methods to avoid iterating by index.
</blockquote>
It was one of two method from original i left untouched. Now it is replaced with <i>hasTrailingWhitespace</i>.<br>
I preferred to retain "allowLineBreaks" there, iteration avoided without loss of functionality.
</p>

<p>
5) 
<blockquote>
I like your report, please put a title to first column "Anchor"
</blockquote>
Done
<blockquote>
I would love to incorporate your tool in Checkstyle testing. 
</blockquote>
It is a hastily assembled pile of dirty code i wrote for a single purpose. I can focus on improving the tool only after mid-november.
</p>
<p>

7)
<blockquote> 
please validate all difference yourself and confirm on next generation that all changes are valid. That will make our cooperation quicker.
</blockquote>
<b>Will do tomorrow</b>
</p>

<p>
8)
<blockquote> 
DetailAST candidate;<br><br>

please rename to "result" in all use-cases.
</blockquote> 
Done.
</p>

<p>
9)
<blockquote>
 final int parentType = parent.getType();<br>
// generics<br>
if (parentType == TokenTypes.TYPE_ARGUMENT) {<br><br>

Why not to use a SWITCH there ?
</blockquote>
<i>Done</i>.
</p>

<p>
10)
<blockquote>
throw new IllegalStateException("unexpected ast syntax" + parent);<br><br>

by what Input file you have unexpected syntax ? Checkstyle do not work reliably on non-compilable sources by design.
</blockquote>
There is a point where i have to map execution branches onto token types. I'm sure that ast structure will change with future java releases, and new token types will be added, i don't want to "swallow" some new token type in "default" branch. I expect that anywhere else meeting unexpected token type will cause npe, but here i want to retain the exception.
</p>
<p>
11)
<blockquote>
private static DetailAST getArrayDeclaratorParent(DetailAST ast) {<br><br>

rename to getFirstNonArrayDeclaratorParent.
</blockquote>
Done.
</p>
<p>
12)
<blockquote>
Please confirm that after reverting CS-SI/Orekit@05c4911 checkstyle validation passed without any problem on this project. Please send revert of this commit to our fork - https://github.com/checkstyle/Orekit/ to prove that.
</blockquote>
<b>Will do tomorrow</b>
</p>
<p>
13)
<blockquote>
after you rebase your code over our HEAD you should got build failure at http://attatrol.github.io/site_result/checkstyle_merged.html#nput.1.54 , so please provide fix for this case by separate commit of the same PR.
</blockquote>
Failed to reproduce your build failure:<br>
here is <a href="http://attatrol.github.io/site_result2/checkstyle_merged.html">report before rebase</a>,<br>
here is <a href="http://attatrol.github.io/site_result2_rebased/checkstyle_merged.html">report after rebase</a>,
both identical<br>
you link is <a href = "http://attatrol.github.io/site_result2_rebase/checkstyle_merged.html#nput.1.56"> http://attatrol.github.io/site_result2_rebase/checkstyle_merged.html#nput.1.56</a> in the newest report<br>
(they differ from <a href="http://attatrol.github.io/site_result/checkstyle_merged.html">the first report.</a>because i have cleaned tester local repository during os reinstallation and got newer sources).
</p>
<p>
@rnveach 
<blockquote>
please don't use full file path, instead use "getNonCompilablePath" method.
At top of class override "getNonCompilablePath" and follow same pattern as "getPath"s implementation.
</blockquote>
Done

romani commented:
> Failed to reproduce your build failure:

Sorry I mislead you, no build violations are expected, your Check found violation in Input file -  checkstyle/src/it/resources/com/google/checkstyle/test/chapter4formatting/rule4832nocstylearray/ArrayTypeStyleInput.java, we ignore input files during Checkstyle execution owr its own sources.

romani commented:
1)
> if (hasTrailingWhitespace(ast, whitespaceColumnNo,
            whitespaceLineNo)) {

make it single line.

2)
>     boolean hasTrailingWhitespace(DetailAST ast,
        int whitespaceColumnNo, int whitespaceLineNo) {

single exit point from method.
And put it below to follow order of method as they appear in code.

3)
> whitespaceFollowedAst = ast;

still not clear , please explain.

4)
>```
+    @Test(expected = IllegalStateException.class)
+    public void testVisitTokenSwitchReflection() throws Exception {
```

please avoid usage of that annotation, you have to check that message is the same. Right now you are ok on any IllegalStateException event not yours.

5)
>  Finds previous node by text order for an array declarator,
> private static DetailAST caseTypeProcessing(DetailAST ast, DetailAST parent) {

So processing or find ? Please rename method.

attatrol commented:
@romani
7)
<blockquote>
please validate all difference yourself and confirm on next generation that all changes are valid. That will make our cooperation quicker.
</blockquote>
Done.
</p>
<p>
12)
<blockquote>
Please confirm that after reverting CS-SI/Orekit@05c4911 checkstyle validation passed without any problem on this project. Please send revert of this commit to our fork - https://github.com/checkstyle/Orekit/ to prove that. 
</blockquote>
Reverted commit you mentioned, got<br>
<a href=https://gist.github.com/attatrol/b1750cf1c96bfaafde95>report before reverting the commit</a>,<br>
<a href=https://gist.github.com/attatrol/41b83f30245643228a44>report after reverting the commit</a><br>
using online text diffchecker <a herf=https://www.diffchecker.com/diff>https://www.diffchecker.com/diff</a><br>
got <a href=https://gist.github.com/attatrol/4ca4b44703805cfe0160>diff from reports</a>,<br>
it seems warnings mentioned in diff returned to code with reverting CS-SI/Orekit@05c4911.
All of them unrelated to the issue, they are fixed in separate commit.<br>
all warnings fixed in separate commit.
PR for checkstyle/Orekit prepared:<a href=https://github.com/checkstyle/Orekit/pull/13>https://github.com/checkstyle/Orekit/pull/13</a>
<p>
</p>
1)
<blockquote>
if (hasTrailingWhitespace(ast, whitespaceColumnNo,
whitespaceLineNo)) {

make it single line.
</blockquote>
Done.

2)

<blockquote>
boolean hasTrailingWhitespace(DetailAST ast,
   int whitespaceColumnNo, int whitespaceLineNo) {

single exit point from method.
And put it below to follow order of method as they appear in code.
</blockquote>
Done. Why dont you enforce this requirement with ReturnCount?
</p>
<p>

3)
<blockquote>
whitespaceFollowedAst = ast;

still not clear, please explain.
</blockquote>
Well, this part is difficult to expain. First of all, this piece of code exists in original check (in the <i>visitToken</i>), thus it is a part of the original behaviour of the check.<br>
During the issue solving i was guided by <a href=https://github.com/checkstyle/checkstyle/blob/master/src/test/resources/com/puppycrawl/tools/checkstyle/checks/whitespace/InputNoWhitespaceAfterArrayDeclarations.java> "InputNoWhitespaceAfterArrayDeclarations"</a>
test cases of the original check.<br>
As it is described there, actually it is checked that there is no whitespace <i>before</i> <code>ARRAY_DECLARATOR</code> node, not after it, thats why we need to process this type of node separately (as well as similar <code>INDEX_OP</code> node). For all other node types it is checked that there is no whitespace after them: <code>whitespaceFollowedAst = ast;</code>.<br>
I clarified this fact in javadoc and xdoc (today).<br>
There is no other whitespace check that takes <code>ARRAY_DECLARATOR</code> or <code>INDEX_OP</code> node as an argument.<br>
Again, this behaviour is inherited from previous version, i think that <code>ARRAY_DECLARATOR</code> & <code>INDEX_OP</code> nodes processing should be moved either in <i> InputNoWhitespaceBefore</i> or a new check, but this will damage backward compatibility.
</p>
<p>
4)
<blockquote>
@Test(expected = IllegalStateException.class)
public void testVisitTokenSwitchReflection() throws Exception {

please avoid usage of that annotation, you have to check that message is the same. Right now you are ok on any IllegalStateException event not yours.
</blockquote>
Done.
</p>
<p>
5)
<blockquote>
Finds previous node by text order for an array declarator,
private static DetailAST caseTypeProcessing(DetailAST ast, DetailAST parent) {

So processing or find ? Please rename method.
</blockquote>
Done. Now it is <i>getPreviousNodeWithParentTypeType</i>, but it sounds really bad. It was extracted from long method <i>getArrayDeclaratorPreviousElement</i> so it is difficult to describe its purpose without clumsiness.
</p>

13)
<blockquote>
after you rebase your code over our HEAD you should got build failure at http://attatrol.github.io/site_result/checkstyle_merged.html#nput.1.54 , so please provide fix for this case by separate commit of the same PR. 
</blockquote>
I'll delete that whitespace and commit when you tell me there is no new mistakes.

romani commented:
>  Why dont you enforce this requirement with ReturnCount?

We still have a lot of cases that are hard to refactor, same cases are not reasonable to refactor as readabilit is decreased. We have more advanced Check implementation in sevntu project that might reconcile such problems.

> There is no other whitespace check that takes ARRAY_DECLARATOR or INDEX_OP node as an argument.

Strange code always point to the problem :) . 
Please create issue on this, we need to be consistent, if name is defining "After" it has to mean "after", and nothing else !! That might explain a strange behaviour at https://github.com/checkstyle/Orekit/pull/13#issuecomment-152022527 , please confirm.

>  but this will damage backward compatibility. 

It is damage in behaviour compatibility, we do this all the time. It is ok. Behaviour predictability in much more valuable.

> I'll delete that whitespace and commit when you tell me there is no new mistakes. 

please do not touch Inputs, Inputs should stay as ugly as they could.

>getPreviousNodeWithParentTypeType

getPreviousNodeWithParentOfTypeAst

attatrol commented:
<blockquotes>
That might explain a strange behaviour at checkstyle/Orekit#13 (comment) , please confirm.
</blockquotes>
I confirm we wrote about the same thing.  I have added <a href=https://github.com/checkstyle/checkstyle/issues/2465>new issue</a>.<br>
<blockquote>
I'll delete that whitespace and commit when you tell me there is no new mistakes.
please do not touch Inputs, Inputs should stay as ugly as they could.
</blockquote>
Sorry, i still dont understand what you want in a separate commit.<br>
<blockquote>
getPreviousNodeWithParentOfTypeAst
</blockquote>
Done.

romani commented:
> Sorry, i still dont understand what you want in a separate commit.

Do not do any changes for Input.

attatrol commented:
@romani
<blockquote>
Do not do any changes for Input.
</blockquote>
So no additional commit needed?

romani commented:
> So no additional commit needed?

yes.

romani commented:
I am done with code review and test validation, ready to merge.
@attatrol , thanks a lot !!

@mkordas , do you want to review code before a merge ?

rnveach commented:
That is all I have to add.

attatrol commented:
@rnveach 
<blockquote>
missing ".checks." as compilable inputs should be located in the same folder structure as their package.
</blockquote>
Done.

romani commented:
merged as FF, thanks a lot.

Please contact me to discuss introduction of your tool our design process, your tool is in HIGH demand. We need README file in your repo to let us know how to work with it.

@attatrol , I just noticed that you do PR from your master, I highly DO NOT recommend to do this. To simplify rebase procedure and be able to work on few issues in parallel, please always do changes in topic branches and do PRs to our master from branches.
